r/src/bpcells-cpp/matrixUtils/Pseudobulk.cpp

```diff
@@ -0,0 +1,112 @@
+// Args:
+// - mat: Matrix to compute variance from.
+// - cell_groups: Mapping of columns to groups.
+// - method: method to use for calculating pseudobulk stats.
+// - transpose: Whether the matrix is transposed. If false, groups columns and calculates stats per-row. If true, groups rows and calculates stats per-column.
+// Returns:
+// - PseudobulkStats struct containing each of the requested pseudobulk aggregations, in the shape of (num_features, num_groups).
+template <typename T>
+PseudobulkStats pseudobulk_matrix(std::unique_ptr<MatrixLoader<T>> &&mat,
+                                  const std::vector<uint32_t>& cell_groups,
+                                  PseudobulkStatsMethod method,
+                                  bool transpose,
+                                  std::atomic<bool> *user_interrupt) {
+    MatrixIterator<T> it(std::move(mat));
+    std::vector<double> group_count_vec;
+    struct PseudobulkStats res;
+    for (auto& cell_group : cell_groups) {
+        if (cell_group >= group_count_vec.size()) {
+            group_count_vec.resize(cell_group + 1);
+        }
+        group_count_vec[cell_group]++;
+    }
+    Eigen::Map<Eigen::Array<double, 1, Eigen::Dynamic>> group_count(group_count_vec.data(), group_count_vec.size());
+    // If transposed, also keep matrix transposed with groups as rows for cache-friendliness
+    uint32_t num_rows = transpose ? group_count_vec.size() : it.rows();
+    uint32_t num_cols = transpose ? it.cols() : group_count_vec.size();
+    if ((method & PseudobulkStatsMethod::NonZeros) == PseudobulkStatsMethod::NonZeros) res.non_zeros = Eigen::ArrayXXd::Zero(num_rows, num_cols);
+    if ((method & PseudobulkStatsMethod::Sum) == PseudobulkStatsMethod::Sum) res.sum = Eigen::ArrayXXd::Zero(num_rows, num_cols);
+    if ((method & PseudobulkStatsMethod::Mean) == PseudobulkStatsMethod::Mean) res.mean = Eigen::ArrayXXd::Zero(num_rows, num_cols);
+    if ((method & PseudobulkStatsMethod::Variance)== PseudobulkStatsMethod::Variance) res.var = Eigen::ArrayXXd::Zero(num_rows, num_cols);
+    // Use a one pass strategy that can do non-zero, mean, and variance at the same time
+    while (it.nextCol()) {
+        const uint32_t col = it.currentCol();
+        if (user_interrupt != NULL && *user_interrupt) return res;
+        while (it.load()) {
+            const uint32_t *row_data = it.rowData();
+            const T *val_data = it.valData();
+            const uint32_t count = it.capacity();
+            // Do transpose logic to calculate row_idx, col_idx out of loop, to avoid repeating conditional checks
+            if (transpose) {
+                // Doing variance calculation also does mean and non-zero count, so specific order is required to make sure we don't double count
+                if ((method & PseudobulkStatsMethod::Variance) == PseudobulkStatsMethod::Variance) {
```

Comment:
I reordered calculation so it is Variance -> Mean -> NonZeros. This allwoed me to remove some conditionals.